include/rocksdb/table.h

```diff
@@ -51,6 +51,55 @@ enum ChecksumType : char {
   kxxHash64 = 0x3,
 };
 
+// `PinningTier` is used to specify which tier of block-based tables should
+// be affected by a block cache pinning setting (see
+// `MetadataCachePinningOptions` below).
+enum class PinningTier {
+  // For compatibility, this value specifies to fallback to the behavior
+  // indicated by the deprecated options,
+  // `pin_l0_filter_and_index_blocks_in_cache` and
+  // `pin_top_level_index_and_filter`.
+  kFallback,
+
+  // This tier contains no block-based tables.
+  kNone,
+
+  // This tier contains block-based tables that may have originated from a
+  // memtable flush. In particular, it includes tables from L0 that are smaller
+  // than 1.5 times the current `write_buffer_size`. Note these criteria imply
+  // it can include intra-L0 compaction outputs and ingested files, as long as
+  // they are not abnormally large for L0.
+  kMaybeFlushed,
+
+  // This tier contains all block-based tables.
+  kAll,
+};
+
+// `MetadataCachePinningOptions` contains members indicating the desired block
+// cache pinning behavior for the different categories of metadata blocks.
+struct MetadataCachePinningOptions {
+  // The tier of block-based tables whose top-level index into metadata
+  // partitions will be pinned. Currently indexes and filters may be
+  // partitioned.
+  //
+  // Note `cache_index_and_filter_blocks` must be true for this option to have
+  // any effect. Otherwise any top-level index into metadata partitions would be
+  // held in table reader memory, outside the block cache.
+  PinningTier top_level_index_pinning = PinningTier::kFallback;
+
+  // The tier of block-based tables whose metadata partitions will be pinned.
```

Comment:
I think a potential fallacy here is "why partition metadata if it's going to be pinned?"  Especially for filters, the CPU hit from key-order partitioning is paid even if the partitions are pinned. I don't think there's much of a case for "what about trivial moves? what about re-opening with different options?" etc. Why not instead skip partitioning in the cases where pinning would be used?

Reply:
Good point. For trivial move I was surprised that it might not even reopen the table reader, so the different settings of the lower level might not even be applied until the DB reopens. So I agree that is not a strong case. The configuration you mentioned could get even weirder by setting `top_level_index_pinning == PinningTier::kNone && partition_pinning >= PinningTier::kMaybeFlushed` (the testing shows such a config does work though). My only guess at a valid use case of `partition_pinning >= PinningTier::kMaybeFlushed` would be for people who really don't like large allocations, maybe due to fragmentation.

Do you think this is needed to proceed with this PR? Pinning partitions was already partially supported by the existing option, `pin_l0_filter_and_index_blocks_in_cache`.

Reply:
I guess we can punt on that improvement.